### ISSUE TITLE
SelectMany<T>() and Concat<T>()

### ISSUE THREAD

**Issue by TonyValenti** [dotnet/runtime#54220]
## Background and Motivation
When working with different APIs, often times it increases code readability to aggregate results and then linearize them.  For example, I could write the code:
```
Command.NameInfo.OnlineAccounts = new List<OnlineAccount>();
Command.NameInfo.OnlineAccounts.AddRange(Input.WebAddresses.Select(x => CreateWebAddress(x)));
Command.NameInfo.OnlineAccounts.AddRange(Input.EmailAddresses.Select(x => CreateEmailAddress(x)));
```
or I could write:
```
            Command.NameInfo.OnlineAccounts = new[] {
                Input.WebAddresses.Select(x => CreateWebAddress(x)),
                Input.EmailAddresses.Select(x => CreateEmailAddress(x))
            }.SelectMany().ToList();
```
The second is much more clear and concise.

## Proposed API
```
public static IEnumerable<T> SelectMany<T>(this IEnumerable<IEnumerable<T>> This) {
            return This.SelectMany(x => x);
        }

        public static IEnumerable<T> Concat<T>(this IEnumerable<IEnumerable<T>> This) {
            var ret = Enumerable.Empty<T>();

            foreach (var item in This) {
                if(item == null) {
                    throw new NullReferenceException();
                }

                if(ret == Enumerable.Empty<T>()) {
                    ret = item;
                } else {
                    ret = Enumerable.Concat(ret, item);
                }

            }

            return ret;
        }
```

## Usage Examples
```
            Command.NameInfo.OnlineAccounts = new[] {
                Input.WebAddresses.Select(x => CreateWebAddress(x)),
                Input.EmailAddresses.Select(x => CreateEmailAddress(x))
            }.SelectMany().ToList();
```

## Alternative Designs
Just use ```SelectMany(x => x)```

## Risks
None

**Comment by eiriktsarpalis** [comment]
I'm assuming `SelectMany()` and `Concat()` are just alternative names for the same proposed method? Why not just stick with one? I personally would prefer exposing `Concat()` as an overload, since F# uses the same name for [its equivalent method](https://fsharp.github.io/fsharp-core-docs/reference/fsharp-collections-seqmodule.html#concat).

**Comment by svick** [comment]
Note that another way to write this code is to use the existing overload of `Concat`:

```c#
Command.NameInfo.OnlineAccounts = 
    Input.WebAddresses.Select(x => CreateWebAddress(x))
        .Concat(Input.EmailAddresses.Select(x => CreateEmailAddress(x)))
        .ToList();
```

Or:

```c#
Command.NameInfo.OnlineAccounts = Enumerable.Concat(
    Input.WebAddresses.Select(x => CreateWebAddress(x)),
    Input.EmailAddresses.Select(x => CreateEmailAddress(x))
).ToList();
```

But you do have to choose between the fluent assymetric form and the non-fluent symmetric form. So I think the proposed overload of `Concat` would still be a good addition.

**Comment by eiriktsarpalis** [comment]
Agreed, @TonyValenti if you could update the OP following the API proposal template we can consider this proposal for .NET 7.

**Comment by danielchalmers** [comment]
I often make a `Flatten<T>(this IEnumerable<T> l)=>l.SelectMany(x=>x)` extension so I would love to see this in the BCL!

**Comment by eiriktsarpalis** [comment]
@danielchalmers please consider upvoting the original post, it provides a good signal on whether a particular proposal is popular!

**Comment by eiriktsarpalis** [comment]
Hi @TonyValenti, before I can mark this ready for API review would it be possible to update the original post so that it follows our [API review template](https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md#steps)?

**Comment by eiriktsarpalis** [comment]
A few issues with the original text I would like to call out:

* Proposal contains two different methods for the same functionality: `SelectMany` and `Concat`. One needs to be removed, I would recommend going for `Concat`.
* The proposed API is missing namespace and declaring class for the proposed methods.
* The proposed API contains implementation, which is not needed from the API review perspective.

The API review document contains links to a few API proposal, if you're looking for real-world examples.

**Comment by TonyValenti** [comment]
@eiriktsarpalis Thanks!  I just updated it.  I left the implementation just because I felt that added more clarity.  I also added a params overload for Concat that will make joining multiple enumerables easier.

**Comment by eiriktsarpalis** [comment]
@TonyValenti I've updated your proposal slightly to better match the template. The reason we want this is to allow the API review committee quickly read and understand what is being proposed.

**Comment by bartonjs** [comment]
> Did we discuss params Span at all? Assuming that shows up in C# 11, would we still use params arrays here, would we add both, or would we say as a general rule we prefer spans moving forward?

We did.  Partly it'll end up depending on what it looks like for those methods to be called by VS/F#.  If it Just Works, then @terrajobst suggested we'd go back and look at new-in-7 API that we could remove the redundancies from.  If it doesn't just work, then we'll probably add them in parallel anywhere other than low-level types.

**Comment by eiriktsarpalis** [comment]
Changed to `api-needs-work` due to feedback in #61230. cc @stephentoub

**Comment by TahirAhmadov** [comment]
@bartonjs did you consider the solution I proposed [here](https://github.com/dotnet/csharplang/discussions/5477#discussioncomment-1708489) ?

**Comment by LeaFrock** [comment]
No progress for .NET 8? Does `IQueryable` block it?

**Comment by JBrown526** [comment]
It would be nice to have a `Flatten()` or parameterless `SelectMany()` implementation at least, as the current usage of `nestedCollections`.SelectMany(x => x);` never seems particularly intuitive or meaningful for me

**Comment by Thaina** [comment]
I would like to add my voice to `Flatten()` or parameterless `SelectMany()` too. I think the old BCL was not support this because it lack contravariant feature? If so we now have contravariant it should be allowed?
